### ISSUE TITLE
Extend nn.functional.softmax for arbitrary dimensions

### ISSUE THREAD

**Issue by Parskatt** [pytorch/pytorch#34874]
## 🚀 Feature

Allow dim in softmax to be: dim (python:int or tuple of python:ints).
## Motivation

This would allow calculation of softmax over arbitrary dimensions, which is useful in some contexts.

Example of what I'm currently doing in a project(not numerically stable):

```python

def softmax(x,dim):
    exp_x = torch.exp(x)
    return exp_x/torch.sum(exp_x,dim=dim,keepdim=True)
```
Many pytorch functions already allow a Union of int and tuples for dimensions. 
Extending this to softmax would be in line with other functions (e.g. torch.sum etc), so I think there is no ambiguity with allowing arbitrary dimensions for normalisation.

## Alternatives
None that I have found

**Comment by HopefulRational** [comment]
Hi @Parskatt. I think this can be done by using `permute` and  `reshape` to merge those dimensions and then using `nn.functional.softmax` for that single dimension and again using the `permute` and `reshape` again to get to right dimensions.
I came across this and did it that way while implementing a paper that uses softmax_3D. Following code may be helpful for better understanding : 
```
# tnsr : tensor (dim0, dim1, dim2, dim3, dim4, dim5)
# softmax of tnsr along (1,2,4)
tmp = tnsr.permute(0,3,5,1,2,4).contiguous().reshape(dim0, dim3, dim5, dim1*dim2*dim4)
res = nn.functional.softmax(tmp, dim=-1)
res = res.reshape(dim0, dim3, dim5, dim1, dim2, dim4).permute(0,3,4,1,5,2).contiguous()
```
Hope this helps :-)

**Comment by Parskatt** [comment]
Yeah, but it looks messy ;)

**Comment by mrshenli** [comment]
Adding this to triage review to discuss how to proceed on this feature request.

**Comment by gchanan** [comment]
I think we should always accept these kinds of requests, but they shouldn't be "high priority."

**Comment by ezyang** [comment]
I agree that permuting and then unpermuting is probably the easiest way to implement it.

**Comment by yf225** [comment]
@Parskatt We would love to accept PRs on this feature, thanks!
